Display notice when user has unactivated product license keys

elliottprogrammer opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->

This PR displays a ""You have an inactive product. Activate now" notice in the dashboard if the user has unactivated product license keys available.

Note: This PR is a "re-do", based off https://github.com/Automattic/jetpack/pull/21426 which was closed because of complications with failing build checks. You can reference #21426 to view the original comments.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->

- Register API route (/jetpack/v4/licensing/user/counts)  for fetching user license counts.
- Fetch user license counts and store them in initialState.
- Register API route (/jetpack/v4/licensing/user/activation-notice-dismiss)  for storing activation notice dismissal info.
- Create reducer, actions, and selectors for retrieving and updating user license counts and notice dismissal info in redux state.
- Persists the notice's custom dismissal info in wp_options.
- Displays the notice when the user acquires a new license, or when the user has an available license(s), but has dismissed the notice and it's been over 2 weeks without activating it.
- Removes notice after a license key is activated, or if the notice is dismissed.
- 

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

- Spin up a site using this Jetpack branch. 
- Connect Jetpack to your WordPress.com user account.
- Enable the license-key input box UI on the "My plan" page by adding `add_filter( 'jetpack_licensing_ui_enabled', '__return_true' );` into Jetpack.
- Go to `/wp-admin/?page=jetpack#/dashboard#/my-plan` (aka "My plan" page)
- Verify you **do not see a notice** at the top that says, "You have an inactive product. Activate now." (Assuming you don't have any unattached user-licenses associated to your user account, which you shouldn't...yet..)
- Verify you can see the License Key input field (at the bottom of the "My Plan" section).

Now obtain a 'user' product license key:

- log into your wpcom sandbox, and in file `wp-content/mu-plugins/0-sandbox.php` add: `add_filter( 'jetpack_license_checkout_m2_enabled', '__return_true' );`. (Enabling this flag makes Jetpack checkout issue you a license when completing checkout).
- Open a new command-line window in your sandbox and `tail -f /tmp/php-errors`
- Point `public-api.wordpress.com` to your sandbox IP (in your hosts file).
- Make sure you are proxied and go to https://wordpress.com/checkout/jetpack/jetpack_scan (siteless checkout flow).
- Complete checkout (using credits is fine).
- View your error log, you should see the output: `Jetpack user license was created for subscription_id:17829922, belonging to user_id:78351100 and blog_id:198416894`, but with different id values corresponding to your user and site. And you'll also see, `License key is: jetpack_scan_monthly_oRhIEF4ckyVP6mCuHaD1vYnw5`, but with your unique key.
- Copy the license key from the error log output, and paste it somewhere for reference. We'll need this license key in a later step.
- un-sandbox `public-api.wordpress.com`.

**Now test that the notice appears now that you have been provisioned a detached license**
- Go to your Jetpack site and reload `/wp-admin/?page=jetpack#/dashboard#/my-plan` (aka "My plan" page).
- Verify that you see a notice at the top that says, **"You have an inactive product. Activate now."**

**Now test the notice dismiss behavior:**
**Prerequisite**: You'll need [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd) installed into your browser for this.

- Open your browser developer tools and open Redux DevTools. (you may need to reload the page.) Inspect the state tree and locate `state.licensing.activationNoticeDismiss`
- Click the notification dismiss button (x), verify that the notice disappears. and the values for `last_detached_count` and `last_dismissed_time` are updated.
- Now, to verify that the notice reappears when a new detached license in acquired, in Redux DevTools, dispatch the action below: (replacing the value of `detached` with the value of <last_detached_count + 1>.

```
{
	type: 'JETPACK_LICENSING_USER_LICENSE_COUNTS_UPDATE',
	counts: {
		attached: 0,
		detached: <enter the value of `activationNoticeDismiss.last_detached_count` + 1(plus one) here>,
		revoked: 0,
		not_revoked: 0,
	}
}
```
- Verify the notice appears again after the above action is dispatched (when a **new** detached license is added).
- Now reload the page, reloading the persisted `activationNoticeDismissed` values in state. The notice should currently be dismissed from when you clicked dismiss(x) button before.
- Now we'll dispatch another action, changing the `last_dismissed_time` to over 2 weeks ago: (`2021-10-13T16:59:22+00:00`)
```
{
	type: 'JETPACK_LICENSING_ACTIVATION_NOTICE_DISMISS_UPDATE',
	dismissData: {
		last_detached_count: <the current value in state, do not change>,
		last_dismissed_time: '2021-10-13T16:59:22+00:00' // <--change to this>
	}
}
```
- Verify the notice reappears (because the `last_dismiss_date` is set to over 2 weeks ago).

**Now test the notice disappears when you activate(submit) a license key:**
- First log into the site's `wp-shell`: (If using jurassic.ninja, ssh into the site using the ssh user and pass that is displayed in the Dashboard, then run `wpsh`. If using local docker env, run `jetpack docker wp shell`)
- run command: `update_option('jetpack_licensing_activation_notice_dismiss', array('last_detached_count' => null, 'last_dismissed_time' => null) );` . This will reset the dismiss notice.
- Go to your Jetpack site and reload `/wp-admin/?page=jetpack#/dashboard#/my-plan` (aka "My plan" page). Verify the activation notice is visible.
- Go back to Calypso, https://wordpress.com/purchases/subscriptions/:site, and copy the license key you purchased in the beginning steps.
- Paste the license key into the License Key input box and submit.
- Verify the "You have an inactive product. Activate now." notice **disappears** (Also you should see that the Jetpack Scan product has been provisioned for your site(You may need to refresh the page to see that))


**Screenshot showing the new notice ("You have an inactive product. Activate now"):**

![Markup 2021-10-29 at 07 25 37](https://user-images.githubusercontent.com/11078128/139426809-34f1568a-9d9b-4d16-8be0-e3bea4136e3c.png)

elliottprogrammer commented:
This PR is a "re-do", based off #21426 which I ended up closing because of complications with failing build checks. You can reference #21426 to view the original comments.

Thankfully, I got all the checks passing in this PR and it's now ready for review. 👍

rcanepa reviewed (dismissed):
LGTM.

I went through the new instructions regarding the notification dismissal and everything worked as expected. Great job, Bryan.

The activation of a license key and the notification dismissal when there are no more detached licenses also worked as expected.

elliottprogrammer commented on projects/plugins/boost/changelog/add-user-license-activate-notice-2:
```diff
@@ -0,0 +1,5 @@
+Significance: patch
+Type: changed
+Comment: update e2e
```
Yes, for sure, thank you! Not completely sure how or why I did that. 🤷‍♂️   Removed. :)

elliottprogrammer commented on projects/plugins/jetpack/_inc/client/components/jetpack-notices/user-license-activation.jsx:
```diff
@@ -0,0 +1,101 @@
+} from 'state/licensing';
+import SimpleNotice from 'components/notice';
+
+class UserLicenseActivationNotice extends React.Component {
```
Ahh, Yes definitely. Ok got it, done 👍  thank you.

elliottprogrammer commented on projects/plugins/jetpack/_inc/client/components/jetpack-notices/user-license-activation.jsx:
```diff
@@ -0,0 +1,101 @@
+
+	render() {
+		// TODO: Update this link to point to the user-license activation route.
+		const USER_LICENSE_ACTIVATION_ROUTE = '/wp-admin/admin.php?page=jetpack#/my-plan';
```
Oh yes, makes sense!  TIL!  Thanks!

elliottprogrammer commented on projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -1013,6 +1052,72 @@ public static function get_products( $request ) {
+			? $default['last_detached_count']
+			: $request['last_detached_count'];
+		// Use UTC timezone and convert to ISO8601 format(DateTime::W3C) for best compatibility with JavaScript Date in all browsers.
+		$last_dismissed_time = ( new DateTime( 'NOW', new DateTimeZone( 'UTC' ) ) )->format( DateTime::W3C );
```
Ahh yes, totally makes sense, thank you!
Ok, I created a new permission check function, `user_licensing_permission_check()`, that checks that the user is the connection owner and the user is connected. What do you think? Do you foresee any issues there?

jeherve replied:
> since the time difference calculation is being done client-side in the browser

Ah yes, that's a good point. I suppose if we were to  switch to `wp_timezone()` here, we would then have to switch to using `@wordpress/date` on the client side as well to keep things in sync.

Your approach is probably simpler though :)

elliottprogrammer replied:
> Ah yes, that's a good point. I suppose if we were to switch to wp_timezone() here, we would then have to switch to using @wordpress/date on the client side as well to keep things in sync.

Ohh, Does `@wordpress/date` somehow know and use the site's set timezone? I see in the docs it says. "Defaults to timezone from site.".  Ohh, I didn't know that, before, TIL. Although I still don't understand offhand how the @wordpress/date package would know the site's timezone.. 🤔  

Well anyway, yes if you don't mind, I'll just keep it as it is, for simplicity sake. 🙂

jeherve replied:
It's some of the magic of the WordPress packages :). If you type in `wp.date.__experimentalGetSettings().timezone.string` in your browser console while in the block editor or in the Jetpack Dashboard, you should see it in action.

The package lives here:
https://github.com/WordPress/gutenberg/tree/1b0ff7e8606cde2b71270f7744897eb2819eecd7/packages/date

> Although I still don't understand offhand how the @wordpress/date package would know the site's timezone.. 🤔

The info gets passed down from PHP to JS via `wp_localize_script()`, here:
https://github.com/WordPress/WordPress/blob/aa62c253e0639924ad9453ee88d3b9851ccdff4d/wp-includes/script-loader.php#L349-L353

We use the package in some places in the Jetpack dashboard, but not directly to fetch the timezone so far. Here is an example:
https://github.com/Automattic/jetpack/blob/e13487d0dbbe4c743739b6f1a3e37c143a8c2614/projects/plugins/jetpack/_inc/client/components/product-expiration/index.jsx#L53-L60

> I'll just keep it as it is, for simplicity sake. 🙂

👍

elliottprogrammer commented on projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -723,6 +723,45 @@ public static function register_endpoints() {
+					),
+				),
+			)
+		);
```
Ahh yes, totally makes sense, thank you!
Ok, I created a new permission check function, `user_licensing_permission_check()`, that checks that the user is the connection owner and the user is connected. What do you think? Do you foresee any issues there?

jeherve replied:
>  What do you think? Do you foresee any issues there?

It seems good to me 👍

elliottprogrammer commented on projects/plugins/jetpack/_inc/lib/admin-pages/class-jetpack-redux-state-helper.php:
```diff
@@ -363,6 +365,22 @@ public static function get_purchase_token() {
+			'last_dismissed_time' => null,
+		);
+
+		return get_option( 'jetpack_licensing_activation_notice_dismiss', $default );
```
Absolutely. Yes, I should have recognized this. Awesome, thanks!

elliottprogrammer commented on projects/plugins/jetpack/_inc/lib/admin-pages/class-jetpack-redux-state-helper.php:
```diff
@@ -363,6 +365,22 @@ public static function get_purchase_token() {
+		);
+
+		return get_option( 'jetpack_licensing_activation_notice_dismiss', $default );
+	}
```
Ahh yes, that makes sense, thanks. 👍  Will do.

elliottprogrammer commented:
> I gave this a first look, and left a few comments below. Let me know if you have any questions!

@jeherve, thanks very much for your thoughtful feedback and your time! 🙂  I've addressed all your comments in your first look.  Would you mind taking another look? Thanks so much. :)

jeherve commented on projects/plugins/jetpack/_inc/client/components/jetpack-notices/user-license-activation.jsx:
```diff
@@ -0,0 +1,111 @@
+	const daysNoticeHasBeenDismissed = ( now - lastDismissedTime ) / DAY_IN_MILLISECONDS;
+
+	const trackClick = useCallback( () => {
+		analytics.tracks.recordJetpackClick( 'activate-user-license-notice-click' );
```
Should we track the dismissals as well?

Also, super nitpicks:

I'm wondering if we need to add "click" here, since this is already tracked as a `jetpack_wpa_click` action, and if we should use underscores instead of hyphens? We seem to be using underscores in other places:
https://github.com/Automattic/jetpack/blob/39fa657bfe0aaca179275331752fd99a74ac1c41/projects/plugins/jetpack/_inc/client/at-a-glance/stats/dash-stats-bottom.jsx#L55-L57

elliottprogrammer replied:
> Should we track the dismissals as well?
Yes definitely, good catch, thank you. :)  I added tracks events for the notice 'view', 'click', and 'dismiss' events.

jeherve reviewed (commented):
I gave this another try, here are my findings. The site ID I used in my test is `199469670`.

-------

> Verify you do not see a notice at the top that says, "You have an inactive product. Activate now." (Assuming you don't have any unattached user-licenses associated to your user account, which you shouldn't...yet..)

I do see the notice on all dashboard routes, right after connecting the site to WordPress.com, as soon as I've enabled the UI with the filter:

![image](https://user-images.githubusercontent.com/426388/140227604-004746ab-c164-45f1-9c0e-ff8efcce268e.png)

I should note that I had already created another siteless license in the past in my last test. Is this why this is showing now?

------

Maybe we shouldn't display the notice when you're already on the My Plan page, since clicking on the link won't do anything when you're already there?

-----

> Verify the "You have an inactive product. Activate now." notice disappears (Also you should see that the Jetpack Scan product has been provisioned for your site(You may need to refresh the page to see that))

It doesn't seem to work for me; the notice remains once I add a license key:

<img width="1187" alt="Screenshot 2021-11-04 at 0 17 11" src="https://user-images.githubusercontent.com/426388/140231937-dc6db63c-22bf-4ae9-ab70-d378fedb2934.png">

If I refresh the page I can see that the scan product has been added to my site though:

![image](https://user-images.githubusercontent.com/426388/140232096-3eb988cd-09c8-452f-ba75-025bd4875ae4.png)

------

I'm seeing a linting test fail:
https://github.com/Automattic/jetpack/runs/4098791034?check_suite_focus=true

But I don't think it's related to this PR, I'm seeing it on other PRs as well. @anomiex Do you think it could be linked to your recent eslint package updates?

rcanepa commented:
Jumping in to answer some questions :)

> I should note that I had already created another siteless license in the past in my last test. Is this why this is showing now?

Yes, if you have a license attached to a temporary site (created via the siteless purchase flow) it will be considered as a detached/available license. Thus, it is expected that you see the banner.

> Maybe we shouldn't display the notice when you're already on the My Plan page, since clicking on the link won't do anything when you're already there?

We will introduce a new activation flow that will be mapped to a new path. You can see some of that work in https://github.com/Automattic/jetpack/pull/21551 and our designs in ceKxiVna5wfpmwFBsojrAx-fi-2219%3A43465. Since that has not been merged yet, we can't make the banner point to it, but we will in a follow-up PR.

> > Verify the "You have an inactive product. Activate now." notice disappears (Also you should see that the Jetpack Scan product has been provisioned for your site(You may need to refresh the page to see that))
> 
> It doesn't seem to work for me; the notice remains once I add a license key:

Could it be possible that you have another detached license? What do you see if you inspect `Initial_State.licensing.userCounts` in the app state?

elliottprogrammer commented:
> I should note that I had already created another siteless license in the past in my last test. Is this why this is showing now?

Yes, I suspect you have a detached(unused) license that was created from when you last tested, so yes, it would be expected that you see the notice. I should point out that in a following project ([Milestone 3](p1HpG7-cSp-p2)), users will have a place where they will be able to see and manage all their licenses, so when that happens users will be able to reference the detached license that is causing the notice and they can choose to activate it, if they wish.

Also, a helpful tip when testing anything with licenses:  You can log into your wpcom sandbox, open wp-shell (`wpsh`), and query the `jetpack_licenses` table to view, update, or delete your licenses (with caution though, because active "partner" licenses (`owner_type`="partner" ) are in this table also. But all the "user" licenses (`owner_type`="user") are all only used for testing at this point.):
```
gpr SELECT * FROM jetpack_licenses WHERE owner_id=<your user id> and attached_at IS NULL;
```
So if you wanted, you could find(in the licenses table) the detached license that is causing the your notice to remain active and copy the license key, and then activate it on some site of yours. That would bring you "detached" licenses count back to 0. Or you could even delete the license directly from the table, (and then remove the associated temporary site subscription in store admin), but I would suggest the former.

> Maybe we shouldn't display the notice when you're already on the My Plan page, since clicking on the link won't do anything when you're already there?

Yes, as Renzo already mentioned, the location where user's will submit(activate) their license key will change and not be the "My Plan" page. In this PR we are only temporarily using the license key input box that was already there for Partner licenses, just because it allows us to test the flow. The new license key input box for 'User' licenses is being introduced in #21551.

> > Verify the "You have an inactive product. Activate now." notice disappears (Also you should see that the Jetpack Scan product has been provisioned for your site(You may need to refresh the page to see that))
> 
> It doesn't seem to work for me; the notice remains once I add a license key:

Yes, this I suspect is for the same reason as stated above, where you still have a detached(unused) license that was created from when you first tested a couple days ago.

However this makes me wonder if when the user submits(activates) their license key, maybe we should manually trigger the notification dismiss action in the background. That way the notification will always disappear on this site after activating a license (even if the user has additional unused(detached) licenses), and the notice will only reappear again when the user obtains(purchases) a new license key, or if the user has not activated their detached licenses for over 2 weeks. @rcanepa, what do you think?

rcanepa commented:
> However this makes me wonder if when the user submits(activates) their license key, maybe we should manually trigger the notification dismiss action in the background. That way the notification will always disappear on this site after activating a license (even if the user has additional unused(detached) licenses), and the notice will only reappear again when the user obtains(purchases) a new license key, or if the user has not activated their detached licenses for over 2 weeks. @rcanepa, what do you think?

I agree 100%. Let's disable the notice even if the user has additional detached licenses.

elliottprogrammer commented on projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -1013,6 +1046,72 @@ public static function get_products( $request ) {
+			'last_dismissed_time' => $last_dismissed_time,
+		);
+
+		Jetpack_Options::update_option( 'licensing_activation_notice_dismiss', $notice_data, true );
+		return rest_ensure_response( Jetpack_Options::get_option( 'licensing_activation_notice_dismiss', $default ) );
```
Yes, good call, thank you.

elliottprogrammer commented:
> I agree 100%. Let's disable the notice even if the user has additional detached licenses.

Looking into this, I'm realizing that I don't think we want to call the `updateLicensingActivationNoticeDismiss` action when the user submits a license key here on the "Partner" license key input box.

We'll want to add it into `projects/js-packages/licensing/components/activation-screen/index.jsx` that @cbauerman  created in #21551.

elliottprogrammer commented:
So the status of this PR is that all the essential functionality is in place, working correctly, and ready for Crew re-review. @jeherve 

Then after this PR is merged, we'll need to add into #21551 a function call to `updateLicensingActivationNoticeDismiss` in `projects/js-packages/licensing/components/activation-screen/index.jsx` inside the `.then()` block [here](https://github.com/Automattic/jetpack/pull/21551/files#diff-1e242004e25d280c0dd0f2a4499c4c0336ce4e460e2c24eabeb6397d32fba50bR47). 

Doing this will make sure that the **"You have an inactive product. Activate now."** notice gets dismissed every time the user submits(activates) their license key (even if the user owns additional unattached licenses). The notice will then stay dismissed until either: the user obtains a new license key for a new product, or if the user still has any unattched licenses that have not been attached for over 2 weeks (reminding them that they have an unused product license key waiting to be activated).

jeherve reviewed (approved):
This looks good and tests well for me. I only have one question left, but I'll approve this PR anyway since I think this can be addressed in a follow-up PR.

elliottprogrammer commented on projects/plugins/jetpack/_inc/client/components/jetpack-notices/index.jsx:
```diff
@@ -276,6 +277,7 @@ class JetpackNotices extends React.Component {
 						onDismissClick={ this.props.clearLicensingError }
 					/>
 				) }
+				<UserLicenseActivationNotice />
```
No, actually we want this to apply to everyone. (although the notice will only appear if a user has an unattached "user" license). 

The `jetpack_licensing_ui_enabled` filter is specifically for the licensing UI used for 'Partner' licenses, however we will be using a different licensing UI specifically for "user" (not partner) licenses, that has been created in #21551.  
I only used and enabled the `jetpack_licensing_ui_enabled` filter in the testing instructions to demonstrate what happens when submitting a license key (because it just so happens to work fine for attaching "user" licenses too).

jeherve replied:
Thanks for the explanation!

elliottprogrammer commented:
r234892-wpcom
